File: external/HDF-EOS/hdfeos/samples/ReadLevels.c
#include "mfhdf.h"
#include <math.h>

/* ReadLevels */

main()
{

    intn            status, i, j;
    
    int32           ptfid, n, nt, sz;
    int32           PTid, date, i32;
    float32         rain, temp, conc[4];
    float64         lon, lat, time;
    int32           fldsz, fldlevels[32], nattr;
    int32           fldtype[32], fldorder[32], recs[128];
    
    char           *pntr, *buffer, id[2], desc[16], spc[8];
    char            fldlist[128], attrnames[16];



    /*
     * Open the HDF swath file, "PointFile.hdf".
     */

    ptfid = PTopen("PointFile.hdf", DFACC_READ);


    /* Read Simple Point */
    /* ----------------- */
    PTid = PTattach(ptfid, "Simple Point");

    status = PTlevelinfo(PTid, 0, fldlist, fldtype, fldorder);
    fldsz = PTsizeof(PTid, fldlist, fldlevels);
    n = PTnrecs(PTid, 0);

    buffer = (char *) calloc(n * fldsz, 1);
    for (i=0; i<n; i++)
    {
	recs[i] = i;
    }
    
    status = PTreadlevel(PTid, 0, fldlist, n, recs, buffer);
    pntr = buffer;

    for (i=0; i<n; i++)
    {
	memcpy(&time, pntr, 8);
	pntr += 8;
	memcpy(conc, pntr, 4*4);
	pntr += 4*4;
	memcpy(spc, pntr, 4);
	pntr += 4;
	printf("%12.1f %6.2f %6.2f %6.2f %6.2f %s\n", 
	       time,conc[0],conc[1],conc[2],conc[3],spc);
    }
    
    free(buffer);
    PTdetach(PTid);
    


    /* Read Fixed Buoy Point */
    /* --------------------- */
    PTid = PTattach(ptfid, "FixedBuoy Point");


    /* Read First (0th) Level */
    /* ----------------------- */
    status = PTlevelinfo(PTid, 0, fldlist, fldtype, fldorder);
    fldsz = PTsizeof(PTid, fldlist, fldlevels);

    n = 2;
    recs[0] = 0;
    recs[1] = 2;
    
    buffer = (char *) calloc(n * fldsz, 1);
    pntr = buffer;
    status = PTreadlevel(PTid, 0, fldlist, n, recs, buffer);

    for (i=0; i<n; i++)
    {
	memcpy(desc, pntr, 8);
	pntr += 8;
	memcpy(&lon, pntr, 8);
	pntr += 8;
	memcpy(&lat, pntr, 8);
	pntr += 8;
	memcpy(&date, pntr, 4);
	pntr += 4;
	memcpy(id, pntr, 1);
	pntr += 1;
	printf("%-10s %12.6f %12.6f %10d %-4s\n", desc,lon,lat,date,id);
    }

    free(buffer);


    /* Read Second (1th) Level */
    /* ------------------------ */
    status = PTlevelinfo(PTid, 1, fldlist, fldtype, fldorder);
    fldsz = PTsizeof(PTid, fldlist, fldlevels);
    
    buffer = (char *) calloc(fldsz, 1);
    n = PTnrecs(PTid, 1);

    for (i=0; i<n; i++)
    {
	pntr = buffer;
	recs[0] = i;
	
	status = PTreadlevel(PTid, 1, fldlist, 1, recs, buffer);
	
	memcpy(&time, pntr, 8);
	pntr += 8;
	memcpy(&rain, pntr, 4);
	pntr += 4;
	memcpy(&temp, pntr, 4);
	pntr += 4;
	memcpy(id, pntr, 1);
	pntr += 1;
	printf("%12.1f %6.2f %6.2f %s\n", time,rain,temp,id);
    }

    free(buffer);

    nattr = PTinqattrs(PTid, attrnames);
    status = PTreadattr(PTid, "int32", &i32);
    printf("%d\n", i32);
    status = PTattrinfo(PTid, "int32", &nt, &sz);
    
	
    PTdetach(PTid);

    
    PTclose(ptfid);

    return;
}

File: external/HDF-EOS/hdfeos/samples/readlevels.f
c
c  In this example we will (1) open the "PointFile" HDF file, (2) attach to
c  the points, and (3) read data from each level of the points.
c 
	program readlevels
	

	integer*4         ptfid, ptid, recs(32)
	integer*4         fldtype(32), fldorder(32)

        integer            status, i, pntr
	integer*2          wgt
	integer*4          n, date
	
	real*4             rain, temp, conc(4)
	real*8             lon, lat, time
				
	character         fldlist*256, buffer*100000, id*2
	character          spc*4, desc*8
        character          ctime*8, cconc*16, clon*8, clat*8
	character          cdate*4, crain*4, ctemp*4, cwgt*2
	equivalence        (time,ctime), (conc,cconc), (lon,clon)
	equivalence        (lat,clat), (rain,crain), (temp,ctemp)
	equivalence        (date,cdate), (wgt,cwgt)
			
	integer           ptrdlev
	integer           ptdetach, ptclose, ptlevinfo
	integer*4         ptopen, ptattach, ptnrecs
	
	
	integer DFACC_READ
	parameter (DFACC_READ=1)

	
c
c     Open the HDF swath file, "PointFile.hdf".
c

	ptfid = ptopen("PointFile.hdf", DFACC_READ)

c
c    Read Simple Point
c
	ptid = ptattach(ptfid, "Simple Point")

	status = ptlevinfo(ptid, 0, fldlist, fldtype, fldorder)
	n = ptnrecs(ptid, 0)

	do 5 i=1,n
	   recs(i) = i - 1
 5	continue
	   
	status = ptrdlev(ptid, 0, fldlist, n, recs, buffer)
	pntr = 1

	do 10 i=1,n
	   ctime = buffer(pntr:pntr+8)
	   pntr = pntr + 8
	   cconc = buffer(pntr:pntr+4*4)
	   pntr = pntr + 4*4
	   spc = buffer(pntr:pntr+4)
	   pntr = pntr + 4
	   write(*,*) time, conc(1), conc(2), conc(3), conc(4), spc
10	continue
    
	status = ptdetach(ptid)
    

c
c    Read Fixed Buoy Point
c
	ptid = ptattach(ptfid, "FixedBuoy Point")

c
c       Read First (0th) Level
c
	status = ptlevinfo(ptid, 0, fldlist, fldtype, fldorder)

	n = 2
	recs(1) = 0
	recs(2) = 2
    
	pntr = 1
	status = ptrdlev(ptid, 0, fldlist, n, recs, buffer)

	do 20 i=1,n
	   desc = buffer(pntr:pntr+8)
	   pntr = pntr + 8
	   clon = buffer(pntr:pntr+8)
	   pntr = pntr + 8
	   clat = buffer(pntr:pntr+8)
	   pntr = pntr + 8
	   cdate = buffer(pntr:pntr+4)
	   pntr = pntr + 4
	   id = buffer(pntr:pntr+1)
	   pntr = pntr + 1
	   write(*,*) desc, lon, lat, date, id(1:1)
 20	continue
	
c
c    Read Second (1th) Level
c
	status = ptlevinfo(ptid, 1, fldlist, fldtype, fldorder)
	n = ptnrecs(ptid, 1)


	do 30 i=1,n

	   recs(1) = i - 1
	   status = ptrdlev(ptid, 1, fldlist, 1, recs, buffer)

	   pntr = 1
	   
	   ctime = buffer(pntr:pntr+8)
	   pntr = pntr + 8
	   crain = buffer(pntr:pntr+4)
	   pntr = pntr + 4
	   ctemp = buffer(pntr:pntr+4)
	   pntr = pntr + 4
	   id = buffer(pntr:pntr+1)
	   pntr = pntr + 1
	   write(*,*) time, rain, temp, '  ', id(1:1)
 30	continue	

	status = ptdetach(ptid)
	status = ptclose(ptfid)

	stop
	end
